alpha diversity (maybe others):  Unable to fetch all required data (during a compute)

we still these errors occasionally
[compute/issues/93](https://github.com/VEuPathDB/service-eda-compute/issues/93)

slack thread in mbiome around these 500 on vizes
UD shared by dan, no subsetting, ranked abundance boxplot 500 (see screnshots)
another UD, no subsetting, same viz selections, doesnt throw error.

```
EDAUD_eBjRGbHgDzH

 "name": "FetchClientError",
 "message": "500 Internal Server Error: POST https://microbiomedb.org/eda/apps/abundance/visualizations/boxplot\nTraceid: 00768468a6694e2e998b95165f4bece9\n\n{\"status\":\"server-error\",\"message\":\"Unable to fetch all required data\",\"requestId\":\"6hffxK9Xi4Vd5k0TPde6sH\"}"


Caused by: java.lang.IllegalStateException: Computed row entity ID 'CB-B6-BP-I-8DPI_S3_R1_mergedLanes (Source) does not match expected ID AIR-BLANK_S22_R1_mergedLanes (Source)
	at org.veupathdb.service.eda.merge.core.stream.RootStreamingEntityNode.applyCompute(RootStreamingEntityNode.java:159) ~[service.jar:3.0.0]
	at org.veupathdb.service.eda.merge.core.stream.RootStreamingEntityNode.lambda$next$7(RootStreamingEntityNode.java:133) ~[service.jar:3.0.0]
	at java.util.Optional.ifPresent(Optional.java:178) ~[?:?]
	at org.veupathdb.service.eda.merge.core.stream.RootStreamingEntityNode.next(RootStreamingEntityNode.java:133) ~[service.jar:3.0.0]
	at org.veupathdb.service.eda.merge.core.MergeDataInputStream.tryQueueNextLine(MergeDataInputStream.kt:117) ~[service.jar:3.0.0]
```

<img width="1286" alt="Image" src="https://github.com/user-attachments/assets/87a78e29-5c02-4d78-abb3-0021d5973454" />
<img width="1286" alt="Image" src="https://github.com/user-attachments/assets/c49ff131-7602-472d-8f22-c50e23135b12" />

--- a/src/main/java/org/veupathdb/service/eda/common/plugin/util/PluginUtil.java
+++ b/src/main/java/org/veupathdb/service/eda/common/plugin/util/PluginUtil.java
@@ -230,6 +230,7 @@ public class PluginUtil {
 
     return fileName +
       " <- data.table::fread(" + singleQuote(fileName) +
+      ", sep='\\t'" +
       ", select=c(" + namedTypes + ")" +
       ", na.strings=c(''))";
   }

--- a/src/main/java/org/veupathdb/service/eda/compute/plugins/alphadiv/AlphaDivPlugin.java
+++ b/src/main/java/org/veupathdb/service/eda/compute/plugins/alphadiv/AlphaDivPlugin.java
@@ -23,6 +23,7 @@ import java.util.List;
 public class AlphaDivPlugin extends AbstractPlugin<AlphaDivPluginRequest, AlphaDivComputeConfig> {
 
   private static final String INPUT_DATA = "alpha_div_input";
+  private static final String SAMPLE_DATA = "sample_entity_input";
 
   public AlphaDivPlugin(@NotNull PluginContext<AlphaDivPluginRequest, AlphaDivComputeConfig> context) {
     super(context);
@@ -31,8 +32,12 @@ public class AlphaDivPlugin extends AbstractPlugin<AlphaDivPluginRequest, AlphaD
   @NotNull
   @Override
   public List<StreamSpec> getStreamSpecs() {
-    return List.of(new StreamSpec(INPUT_DATA, getConfig().getCollectionVariable().getEntityId())
-      .addVars(getUtil().getCollectionMembers(getConfig().getCollectionVariable())));
+    String entityId = getConfig().getCollectionVariable().getEntityId();
+    return List.of(
+      new StreamSpec(INPUT_DATA, entityId)
+        .addVars(getUtil().getCollectionMembers(getConfig().getCollectionVariable())),
+      new StreamSpec(SAMPLE_DATA, entityId)  // just ID columns, no extra vars
+    );
   }
 
   @Override
@@ -47,9 +52,13 @@ public class AlphaDivPlugin extends AbstractPlugin<AlphaDivPluginRequest, AlphaD
     EntityDef entity = meta.getEntity(entityId).orElseThrow();
     VariableDef computeEntityIdVarSpec = util.getEntityIdVarSpec(entityId);
     String computeEntityIdColName = util.toColNameOrEmpty(computeEntityIdVarSpec);
+    String shortIdColName = computeEntityIdColName.contains(".")
+      ? computeEntityIdColName.substring(computeEntityIdColName.lastIndexOf('.') + 1)
+      : computeEntityIdColName;
     String method = computeConfig.getAlphaDivMethod().getValue();
     HashMap<String, InputStream> dataStream = new HashMap<>();
     dataStream.put(INPUT_DATA, getWorkspace().openStream(INPUT_DATA));
+    dataStream.put(SAMPLE_DATA, getWorkspace().openStream(SAMPLE_DATA));
     List<VariableDef> idColumns = new ArrayList<>();
     for (EntityDef ancestor : meta.getAncestors(entity)) {
       idColumns.add(ancestor.getIdColumnDef());
@@ -62,6 +71,12 @@ public class AlphaDivPlugin extends AbstractPlugin<AlphaDivPluginRequest, AlphaD
       computeInputVars.addAll(util.getCollectionMembers(computeConfig.getCollectionVariable()));
       computeInputVars.addAll(idColumns);
       connection.voidEval(util.getVoidEvalFreadCommand(INPUT_DATA, computeInputVars));
+      List<VariableSpec> sampleIdVars = new ArrayList<>();
+      sampleIdVars.add(computeEntityIdVarSpec);
+      sampleIdVars.addAll(idColumns);
+      connection.voidEval(util.getVoidEvalFreadCommand(SAMPLE_DATA, sampleIdVars));
+      // Strip entity prefix from all ID column names to match the short names used by the compute output
+      connection.voidEval("data.table::setnames(" + SAMPLE_DATA + ", names(" + SAMPLE_DATA + "), sub('^[^.]+\\\\.', '', names(" + SAMPLE_DATA + ")))");
       List<String> dotNotatedIdColumns = idColumns.stream().map(VariableDef::toDotNotation).toList();
       StringBuilder dotNotatedIdColumnsString = new StringBuilder("c(");
       boolean first = true;
@@ -82,6 +97,15 @@ public class AlphaDivPlugin extends AbstractPlugin<AlphaDivPluginRequest, AlphaD
 
       connection.voidEval("alphaDivDT <- alphaDiv(abundDT, " +
         PluginUtil.singleQuote(method) + ", TRUE)");
+
+      // Left-join results onto full sample list so samples with no abundance data get NA values.
+      connection.voidEval("alphaDivData <- alphaDivDT@data");
+      connection.voidEval("alphaDivData <- merge(" + SAMPLE_DATA + ", alphaDivData, by=names(" + SAMPLE_DATA + "), all.x=TRUE, sort=FALSE)");
+      connection.voidEval("alphaDivData <- alphaDivData[match(" + SAMPLE_DATA + "[[" +
+        PluginUtil.singleQuote(shortIdColName) + "]], alphaDivData[[" +
+        PluginUtil.singleQuote(shortIdColName) + "]])]");
+      connection.voidEval("alphaDivDT@data <- alphaDivData");
+
       String dataCmd = "writeData(alphaDivDT, NULL, TRUE)";
       String metaCmd = "writeMeta(alphaDivDT, NULL, TRUE)";
 

--- a/src/main/java/org/veupathdb/service/eda/compute/plugins/betadiv/BetaDivPlugin.java
+++ b/src/main/java/org/veupathdb/service/eda/compute/plugins/betadiv/BetaDivPlugin.java
@@ -25,6 +25,7 @@ import static org.veupathdb.service.eda.common.plugin.util.PluginUtil.singleQuot
 public class BetaDivPlugin extends AbstractPlugin<BetaDivPluginRequest, BetaDivComputeConfig> {
 
   private static final String INPUT_DATA = "beta_div_input";
+  private static final String SAMPLE_DATA = "sample_entity_input";
 
   public BetaDivPlugin(@NotNull PluginContext<BetaDivPluginRequest, BetaDivComputeConfig> context) {
     super(context);
@@ -33,9 +34,12 @@ public class BetaDivPlugin extends AbstractPlugin<BetaDivPluginRequest, BetaDivC
   @NotNull
   @Override
   public List<StreamSpec> getStreamSpecs() {
-    return List.of(new StreamSpec(INPUT_DATA, getConfig().getCollectionVariable().getEntityId())
-      .addVars(getUtil().getCollectionMembers(getConfig().getCollectionVariable())
-    ));
+    String entityId = getConfig().getCollectionVariable().getEntityId();
+    return List.of(
+      new StreamSpec(INPUT_DATA, entityId)
+        .addVars(getUtil().getCollectionMembers(getConfig().getCollectionVariable())),
+      new StreamSpec(SAMPLE_DATA, entityId)  // just ID columns, no extra vars
+    );
   }
 
   @Override
@@ -50,9 +54,13 @@ public class BetaDivPlugin extends AbstractPlugin<BetaDivPluginRequest, BetaDivC
     EntityDef entity = meta.getEntity(entityId).orElseThrow();
     VariableDef computeEntityIdVarSpec = util.getEntityIdVarSpec(entityId);
     String computeEntityIdColName = util.toColNameOrEmpty(computeEntityIdVarSpec);
+    String shortIdColName = computeEntityIdColName.contains(".")
+      ? computeEntityIdColName.substring(computeEntityIdColName.lastIndexOf('.') + 1)
+      : computeEntityIdColName;
     String dissimilarityMethod = computeConfig.getBetaDivDissimilarityMethod().getValue();
     HashMap<String, InputStream> dataStream = new HashMap<>();
     dataStream.put(INPUT_DATA, getWorkspace().openStream(INPUT_DATA));
+    dataStream.put(SAMPLE_DATA, getWorkspace().openStream(SAMPLE_DATA));
     List<VariableDef> idColumns = new ArrayList<>();
     for (EntityDef ancestor : meta.getAncestors(entity)) {
       idColumns.add(ancestor.getIdColumnDef());
@@ -66,6 +74,12 @@ public class BetaDivPlugin extends AbstractPlugin<BetaDivPluginRequest, BetaDivC
       computeInputVars.addAll(idColumns);
       connection.voidEval(util.getVoidEvalFreadCommand(INPUT_DATA, computeInputVars));
       computeInputVars.clear();
+      List<VariableSpec> sampleIdVars = new ArrayList<>();
+      sampleIdVars.add(computeEntityIdVarSpec);
+      sampleIdVars.addAll(idColumns);
+      connection.voidEval(util.getVoidEvalFreadCommand(SAMPLE_DATA, sampleIdVars));
+      // Strip entity prefix from all ID column names to match the short names used by the compute output
+      connection.voidEval("data.table::setnames(" + SAMPLE_DATA + ", names(" + SAMPLE_DATA + "), sub('^[^.]+\\\\.', '', names(" + SAMPLE_DATA + ")))");
       List<String> dotNotatedIdColumns = idColumns.stream().map(VariableDef::toDotNotation).toList();
       StringBuilder dotNotatedIdColumnsString = new StringBuilder("c(");
       boolean first = true;
@@ -85,6 +99,14 @@ public class BetaDivPlugin extends AbstractPlugin<BetaDivPluginRequest, BetaDivC
                                                                           ",imputeZero=TRUE)");
       connection.voidEval("betaDivDT <- betaDiv(abundDT, " +
                                                 singleQuote(dissimilarityMethod) + ")");
+
+      // Left-join results onto full sample list so samples with no abundance data get NA values.
+      connection.voidEval("betaDivData <- betaDivDT@data");
+      connection.voidEval("betaDivData <- merge(" + SAMPLE_DATA + ", betaDivData, by=names(" + SAMPLE_DATA + "), all.x=TRUE, sort=FALSE)");
+      connection.voidEval("betaDivData <- betaDivData[match(" + SAMPLE_DATA + "[[" +
+        singleQuote(shortIdColName) + "]], betaDivData[[" +
+        singleQuote(shortIdColName) + "]])]");
+      connection.voidEval("betaDivDT@data <- betaDivData");
 
       String dataCmd = "writeData(betaDivDT, NULL, TRUE)";
       String metaCmd = "writeMeta(betaDivDT, NULL, TRUE)";

--- a/src/main/java/org/veupathdb/service/eda/compute/plugins/rankedabundance/RankedAbundancePlugin.java
+++ b/src/main/java/org/veupathdb/service/eda/compute/plugins/rankedabundance/RankedAbundancePlugin.java
@@ -22,6 +22,7 @@ import java.util.List;
 public class RankedAbundancePlugin extends AbstractPlugin<RankedAbundancePluginRequest, RankedAbundanceComputeConfig> {
 
   private static final String INPUT_DATA = "ranked_abundance_input";
+  private static final String SAMPLE_DATA = "sample_entity_input";
 
   public RankedAbundancePlugin(@NotNull PluginContext<RankedAbundancePluginRequest, RankedAbundanceComputeConfig> context) {
     super(context);
@@ -30,9 +31,12 @@ public class RankedAbundancePlugin extends AbstractPlugin<RankedAbundancePluginR
   @NotNull
   @Override
   public List<StreamSpec> getStreamSpecs() {
-    return List.of(new StreamSpec(INPUT_DATA, getConfig().getCollectionVariable().getEntityId())
-        .addVars(getUtil().getCollectionMembers(getConfig().getCollectionVariable())
-      ));
+    String entityId = getConfig().getCollectionVariable().getEntityId();
+    return List.of(
+      new StreamSpec(INPUT_DATA, entityId)
+        .addVars(getUtil().getCollectionMembers(getConfig().getCollectionVariable())),
+      new StreamSpec(SAMPLE_DATA, entityId)  // just ID columns, no extra vars
+    );
   }
 
   @Override
@@ -47,9 +51,13 @@ public class RankedAbundancePlugin extends AbstractPlugin<RankedAbundancePluginR
     EntityDef entity = meta.getEntity(entityId).orElseThrow();
     VariableDef computeEntityIdVarSpec = util.getEntityIdVarSpec(entityId);
     String computeEntityIdColName = util.toColNameOrEmpty(computeEntityIdVarSpec);
+    String shortIdColName = computeEntityIdColName.contains(".")
+      ? computeEntityIdColName.substring(computeEntityIdColName.lastIndexOf('.') + 1)
+      : computeEntityIdColName;
     String method = computeConfig.getRankingMethod().getValue();
     HashMap<String, InputStream> dataStream = new HashMap<>();
     dataStream.put(INPUT_DATA, getWorkspace().openStream(INPUT_DATA));
+    dataStream.put(SAMPLE_DATA, getWorkspace().openStream(SAMPLE_DATA));
     List<VariableDef> idColumns = new ArrayList<>();
     for (EntityDef ancestor : meta.getAncestors(entity)) {
       idColumns.add(ancestor.getIdColumnDef());
@@ -64,6 +72,12 @@ public class RankedAbundancePlugin extends AbstractPlugin<RankedAbundancePluginR
         addAll(idColumns);
       }};
       connection.voidEval(util.getVoidEvalFreadCommand(INPUT_DATA, computeInputVars));
+      List<VariableSpec> sampleIdVars = new ArrayList<>();
+      sampleIdVars.add(computeEntityIdVarSpec);
+      sampleIdVars.addAll(idColumns);
+      connection.voidEval(util.getVoidEvalFreadCommand(SAMPLE_DATA, sampleIdVars));
+      // Strip entity prefix from all ID column names to match the short names used by the compute output
+      connection.voidEval("data.table::setnames(" + SAMPLE_DATA + ", names(" + SAMPLE_DATA + "), sub('^[^.]+\\\\.', '', names(" + SAMPLE_DATA + ")))");
       // TODO make a helper for this i think
       List<String> dotNotatedIdColumns = idColumns.stream().map(VariableDef::toDotNotation).toList();
       StringBuilder dotNotatedIdColumnsString = new StringBuilder("c(");
@@ -84,6 +98,15 @@ public class RankedAbundancePlugin extends AbstractPlugin<RankedAbundancePluginR
                                                                           ",imputeZero=TRUE)");
       connection.voidEval("abundanceDT <- rankedAbundance(abundDT, " +
                                                           PluginUtil.singleQuote(method) + ")");
+
+      // Left-join results onto full sample list so samples with no abundance data get NA values.
+      connection.voidEval("abundanceData <- abundanceDT@data");
+      connection.voidEval("abundanceData <- merge(" + SAMPLE_DATA + ", abundanceData, by=names(" + SAMPLE_DATA + "), all.x=TRUE, sort=FALSE)");
+      connection.voidEval("abundanceData <- abundanceData[match(" + SAMPLE_DATA + "[[" +
+        PluginUtil.singleQuote(shortIdColName) + "]], abundanceData[[" +
+        PluginUtil.singleQuote(shortIdColName) + "]])]");
+      connection.voidEval("abundanceDT@data <- abundanceData");
+
       String dataCmd = "writeData(abundanceDT, NULL, TRUE)";
       String metaCmd = "writeMeta(abundanceDT, NULL, TRUE)";
 

